### PR TITLE
machine: add support to set size of RingBuffer

### PR DESCRIPTION
**Pull request by sago35**
Initially, the size of the RingBuffer used by the UART is 128.

https://github.com/tinygo-org/tinygo/blob/64d048c47c95bc0a70b4cff109afc1f9039705a4/src/machine/buffer.go#L7

This is not a problem in most cases, but it can sometimes be a problem when handling the wifi of the rtl8720dn.
For example, the size of a small HTTP header is about 200 bytes, and I want to process it all in a single transfer.

https://github.com/tinygo-org/drivers/pull/273

In this PR, the size can be set without changing the existing code.
In this case, I used variadic, but if there is a better way, I would like to adopt it.

### CODE REVIEW

File: src/machine/buffer.go [hunk 1]
```diff
@@ -4,32 +4,38 @@ import (
...
+	head     volatile.Register16
+	tail     volatile.Register16
 }
 
 // NewRingBuffer returns a new ring buffer.
-func NewRingBuffer() *RingBuffer {
-	return &RingBuffer{}
+func NewRingBuffer(size ...int) *RingBuffer {
+	sz := defaultRingBufferSize
+	if len(size) > 0 {
```
Comment (ardnew) [inline src/machine/buffer.go:20]: I believe this `RingBuffer` implementation depends on its size being a power of 2 to support filling _all_ slots in the buffer (although you might not notice an issue until that many slots are filled). 

Might either verify `sz` satisfies that condition, or round up to ensure it does.

**Comment by aykevl** [comment]
I do not think this is a good idea. I have intentionally picked an array instead of a slice to avoid extra indirection and thus overhead. If this is an issue, it may be better to add an API to call a callback on an incoming byte instead (somewhat similar to pin interrupts).

>  For example, the size of a small HTTP header is about 200 bytes, and I want to process it all in a single transfer.

Is this just a performance problem or does the code behave incorrectly for larger buffers?

File: src/machine/buffer.go (hunk 1 above)
Comment (sago35) [reply src/machine/buffer.go:20]: I agree.
Surely some sort of check is needed.

**Comment by sago35** [comment]
> it may be better to add an API to call a callback on an incoming byte instead (somewhat similar to pin interrupts).

I'll try it.

> Is this just a performance problem or does the code behave incorrectly for larger buffers?

I use a larger buffer for communication to match the Arduino implementation.
Using a larger buffer makes the implementation simpler.

At first it was important to have a simple implementation, but now it looks like I can refactor it.
I will try to refactor it a bit so that I don't have to resize the RingBuffer.

**Comment by sago35** [comment]
With the following commit, this PR is no longer needed.
Close this PR.

https://github.com/tinygo-org/drivers/pull/287#issuecomment-888284428

**Comment by ardnew** [comment]
> I will try to refactor it a bit so that I don't have to resize the RingBuffer.

You may have already resolved this, but I wanted to suggest simply duplicating the logic of this `RingBuffer` using a larger array (or slice) private to one of the types you are working with.

It is a very small implementation after all, so it might be easier to copy its functionality than refactoring your code to work with this specific type.

**Pull request closed without merging**